ramips: add CUDY WR1000 support

DavideFioravanti opened the pull request:
Cudy WR1200 is an AC1200 AP with 3-port FE and 2 non-detachable antennas

The "...-factory.bin" image is flashable directly through the stock web interface
Once OpenWrt is installed, the "...-sysupgrade.bin" image must be used.

I edited the old JCG factory image tool to make it compatible with >4MB flash sizes.
I added the -m argument to specify a custom MAXSIZE
If the -m argument is not present the tool should works as a normal 4MB flash size (for backward compatibility)

chunkeey commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -17,6 +18,11 @@ define Build/sercom-footer
 	$(call Build/sercom-seal,-f)
 endef
 
+define Build/jcg-header
```
Can you please unify this helper command definition with the one in rt305x.mk's Build/jcg-header and move it to ramips/image/Makefile?

(Please also move the DEVICE_VARS += JCG_MAXSIZE there).

chunkeey commented on target/linux/ramips/dts/WR1000.dts:
```diff
@@ -0,0 +1,142 @@
+&spi0 {
+	status = "okay";
+
+	w25q64@0 {
```
Ideally, the nodes should be named after their function and not after the device.

https://github.com/devicetree-org/devicetree-specification/blob/master/source/devicetree-basics.rst#generic-names-recommendation

(I would recommend flash@0 here)

chunkeey commented on target/linux/ramips/dts/WR1000.dts:
```diff
@@ -0,0 +1,142 @@
+};
+
+&pcie0 {
+	mt76@0,0 {
```
Please follow the device-tree generic node names recommendation

https://github.com/devicetree-org/devicetree-specification/blob/master/source/devicetree-basics.rst#generic-names-recommendation

I think "wifi0,0" would fit perfectly.

chunkeey commented on target/linux/ramips/dts/WR1000.dts:
```diff
@@ -0,0 +1,142 @@
+		led-upgrade = &led_wps;
+	};
+
+	gpio-keys-polled {
```
Please rename this node to "buttons" or "keys" (whatever fits better).

chunkeey commented on target/linux/ramips/dts/WR1000.dts:
```diff
@@ -0,0 +1,142 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2s", "refclk", "wdt", "p4led_an", "p3led_an", "p2led_an", "wled_an";
```
It would be great if you could add a newline after ~70-80 characters on that line.

chunkeey commented on target/linux/ramips/dts/WR1000.dts:
```diff
@@ -0,0 +1,142 @@
+		};
+	};
+
+	gpio-leds {
```
Please rename this node to just "leds".

chunkeey commented on tools/firmware-utils/src/jcgimage.c:
```diff
@@ -316,6 +323,8 @@ main(int argc, char **argv)
 	char *file2 = NULL;
 	char *version = NULL;
+	size_t maxsize = MAXSIZE;
+	char* endptr;
```
please follow the style of "char *file..." and make change this to

    char *endptr;

(scripts/checkpatch.pl catches stuff like this)

chunkeey commented on tools/firmware-utils/src/jcgimage.c:
```diff
@@ -387,8 +400,8 @@ main(int argc, char **argv)
 
-	if (sizeo > MAXSIZE) {
-		errx(1,"payload too large: %zd > %zd\n", sizeo, MAXSIZE);
+	if (sizeo > maxsize) {
```
WARNING: braces {} are not necessary for single statement blocks
    #87: FILE: tools/firmware-utils/src/jcgimage.c:403:
    +       if (sizeo > maxsize) {
    +               errx(1,"payload too large: %zd > %zd\n", sizeo, maxsize);
            }

chunkeey commented on tools/firmware-utils/src/jcgimage.c:
```diff
@@ -387,8 +400,8 @@ main(int argc, char **argv)
-	if (sizeo > MAXSIZE) {
-		errx(1,"payload too large: %zd > %zd\n", sizeo, MAXSIZE);
+	if (sizeo > maxsize) {
+		errx(1,"payload too large: %zd > %zd\n", sizeo, maxsize);
```
ERROR: space required after that ',' (ctx:VxV)
    #88: FILE: tools/firmware-utils/src/jcgimage.c:404:
    +               errx(1,"payload too large: %zd > %zd\n", sizeo, maxsize);
      
Otherwise I think this looks good.

DavideFioravanti commented:
I am sorry for all these naming errors... I started working on this commit months ago and I probably used old dts as base. I will fix all these errors in the following days. Thank you.

DavideFioravanti committed:
firmware-utils: jcgimage: fix style and compiler warnings

Fixed some compilation warnings.

Signed-off-by: Davide Fioravanti <pantanastyle@gmail.com>

DavideFioravanti committed:
firmware-utils: jcgimage: add support for 4MiB+ images

The first version of this script allowed just 4Mb factory images.
With this patch is possible to set the maximum size of the payload.

For an 8Mb flash, the corresponding maxsize is:
8 * 1024 * 1024 - 5 * 64 * 1024 = 8388608 - 327680 = 8060928

If the -m argument is not set, the default maximum size will be
used for backward compatibility.

Signed-off-by: Davide Fioravanti <pantanastyle@gmail.com>

DavideFioravanti committed:
ramips: unify JCG helper command definition

This patch makes the JCG helper command definition available
for every rampis target

Signed-off-by: Davide Fioravanti <pantanastyle@gmail.com>

DavideFioravanti committed:
ramips: add CUDY WR1000 support

Cudy WR1200 is an AC1200 AP with 3-port FE and 2 non-detachable antennas

Specifications:

MT7628 (580 MHz)
64 MB of RAM (DDR2)
8 MB of FLASH
2T2R 2.4 GHz (MT7628)
2T2R 5 GHz (MT7612E)
3x 10/100 Mbps Ethernet (2 LAN + 1 WAN)
2x external, non-detachable antennas (5dbi)
UART header on PCB (57600 8n1)
7x LED, 2x button

Known issues:
Power LED always ON (probably it's connected directly to power)

Flash instructions
------------------
Load the ...-factory.bin image via the stock web interface

Openwrt upgrade instructions
----------------------------
Use the ...-sysupgrade.bin image for future upgrades.

Revert to stock FW
------------------
Warning! This tutorial will work only with the following OEM FW:
	WR1000_EU_92.122.2.4987.201806261618.bin
	WR1000_US_92.122.2.4987.201806261609.bin
If in the future these firmwares will not be available anymore,
you have to find the new XOR key.

1) Download the original FW from the Cudy website.
	(For example WR1000_EU_92.122.2.4987.201806261618.bin)
2) Remove the header
	dd if="WR1000_EU_92.122.2.4987.201806261618.bin" of="WR1000_EU_92.122.2.4987.201806261618.bin.mod" skip=8 bs=64
3) XOR the new file with this key:
	FOR EU: 7B76741E67594351555042461D625F4545514B1B03050208000603020803000D
	FOR US: 7B76741E675943555D5442461D625F454555431F03050208000603060007010C
	If you don't know ho to do it you can use this tool (CHANGE THE XOR KEY ACCORDINGLY!):
	https://gchq.github.io/CyberChef/#recipe=XOR(%7B'option':'Hex','string':''%7D,'',false)
4) Check if bytes from 0x20 to 0x3f are:
	4C 69 6E 75 78 20 4B 65 72 6E 65 6C 20 49 6D 61 67 65 00 00 00 00 00 00 00 00 00 00 00 00 00 00
5) Flash it as a normal OpenWrt firmware upgrade (REMEMBER TO NOT KEEP SETTINGS!).

Signed-off-by: Davide Fioravanti <pantanastyle@gmail.com>

DavideFioravanti commented:
@chunkeey
I have (hopefully) done all my "homework"
1) Fixed the old JCG factory tool removing the unecessary braces
2) Updated the JCG factory tool with the new options
3) Removed the helper function from rt305x.mk
4) Fixed my previous commit with the right labels..

chunkeey commented:
@DavideFioravanti 

I'm mostly "away" during most of the work-week days. 

From what I can see, you went a bit too 160% with your patches, because I made the mistake of not referring you to OpenWrt's stance on coding style when I made my comments.
https://openwrt.org/submitting-patches#style_check_your_changes and
https://openwrt.org/submitting-patches#submitting_patches

You only need to care about the lines you add or change. Since you fixed more styling issues you made it harder for yourself because these too need to be looked after. From what I can tell, they
look okay. My problem now is that you have two patches with the same subject "rampis: Update JCG factory image tool". This will look odd in the one-line git history, because people will think that either the first patch screwed something up, or the committer made a mistake. In my opinion you should give them each their own subject.

For the patch https://github.com/openwrt/openwrt/pull/2062/commits/d63a861c1b12bf092c68bf6393d9ce3ee67a4d05 I would suggest something like "firmware-utils: jcgimage: add support for 4MiB+ images" . This is more in line with the recommendation "to describe what and why the commit changes something".

For the patch https://github.com/openwrt/openwrt/pull/2062/commits/9db19c23657eccf5e7b7ee825aea3045d8ab08c8 you could choose "firmware-utils: jcgimage: fix style and compiler warnings" or "firmware-utils: jcgimage: style and warning clean-up". Whatever fits better in your opinion.

For the future, please consider that the checkpatch.pl can only do so much and sometimes it does something consistently wrong as well. From what I can tell, OpenWrt tries to implement the linux kernel coding-style for code. If you want to take a look, you can find a up-to-date version right here:

https://www.kernel.org/doc/html/v5.1/process/coding-style.html

DavideFioravanti commented:
@chunkeey 
Fixed!

P.s. I am really sorry for over-complicating things. However thank you for your advice and your patience! :) I hope that the followings PR will be accurate the first try! :)

chunkeey commented:
Merged. Thanks.

adrianschmutzler commented:
@DavideFioravanti Hi, I just did an attempt to update Build/jcg-header so JCG_MAXSIZE can be specified in kilobytes. You find it at my staging tree:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

Can you have a look and maybe test it?

DavideFioravanti commented:
@adrianschmutzler Ok, it should work. However, I'll test your patch in the next few days 😉

adrianschmutzler commented:
/data/openwrt/staging_dir/host/bin/jcgimage -v 92.122  -m $((7872 * 1024)) -u /data/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt76x8/tmp/openwrt-ramips-mt76x8-cudy_wr1000-squashfs-factory.bin -o /data/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt76x8/tmp/openwrt-ramips-mt76x8-cudy_wr1000-squashfs-factory.bin.new

Looks good, I think you can save your time for something useful :-)
